domain/network/service/space.go

```diff
@@ -179,25 +244,163 @@ func generateFanSubnetID(subnetNetwork, providerID string) string {
+	stateSpaces, err := s.spaceService.GetAllSpaces(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	spaceNames := set.NewStrings()
+	for _, space := range stateSpaces {
+		s.modelSpaceMap[space.ProviderId] = space
+		spaceNames.Add(string(space.Name))
+	}
+
+	for _, spaceInfo := range providerSpaces {
+		// Check if the space is already in state,
+		// in which case we know its name.
+		var spaceID string
+		stateSpace, ok := s.modelSpaceMap[spaceInfo.ProviderId]
+		if ok {
+			spaceID = stateSpace.ID
+		} else {
+			// The space is new, we need to create a valid name for it in state.
+			// Convert the name into a valid name that is not already in use.
+			spaceName := network.ConvertSpaceName(string(spaceInfo.Name), spaceNames)
+
+			s.logger.Debugf("Adding space %s from provider %s", spaceName, string(spaceInfo.ProviderId))
+			spaceID, err := s.spaceService.AddSpace(
+				ctx,
+				network.SpaceInfo{
+					Name:       network.SpaceName(spaceName),
+					ProviderId: spaceInfo.ProviderId,
+				},
+			)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			spaceNames.Add(spaceName)
+
+			// To ensure that we can remove spaces, we back-fill the new spaces
+			// onto the modelSpaceMap.
+			s.modelSpaceMap[spaceInfo.ProviderId] = network.SpaceInfo{
+				ID:         spaceID.String(),
+				Name:       network.SpaceName(spaceName),
+				ProviderId: spaceInfo.ProviderId,
+			}
+		}
+
+		err = s.spaceService.saveProviderSubnets(ctx, spaceInfo.Subnets, spaceID, fanConfig)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		s.updatedSpaces.Add(spaceInfo.ProviderId)
+	}
+
+	return nil
+}
+
+// DeltaSpaces returns all the spaces that haven't been updated.
+func (s *ProviderSpaces) deltaSpaces() network.IDSet {
+	// Workout the difference between all the current spaces vs what was
+	// actually changed.
+	allStateSpaces := network.MakeIDSet()
+	for providerID := range s.modelSpaceMap {
+		allStateSpaces.Add(providerID)
+	}
+
+	return allStateSpaces.Difference(s.updatedSpaces)
+}
+
+// DeleteSpaces will attempt to delete any unused spaces after a SaveSpaces has
+// been called.
+// If there are no spaces to be deleted, it will exit out early.
+func (s *ProviderSpaces) deleteSpaces(ctx context.Context) ([]string, error) {
+	// Exit early if there is nothing to do.
+	if len(s.modelSpaceMap) == 0 {
+		return nil, nil
+	}
+
+	// Then check if the delta spaces are empty, if it's also empty, exit again.
+	// We do it after modelSpaceMap as we create a types to create this, which
+	// seems pretty wasteful.
+	remnantSpaces := s.deltaSpaces()
+	if len(remnantSpaces) == 0 {
+		return nil, nil
+	}
+
+	// TODO (manadart 2024-01-29): The alpha space ID here is scaffolding and
+	// should be replaced with the configured model default space upon
+	// migrating this logic to Dqlite.
+	defaultEndpointBinding := network.AlphaSpaceId
```

Comment:
I leave this here for now. The idea is to inject the default endpoint binding space from the model config on the places where we recruit the Reload spaces functionality.